src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs

```diff
@@ -0,0 +1,229 @@
+                }
+            }
+            return value;
+        }
+
+        private string HtmlEncodeImpl(string input, int idxOfFirstCharWhichRequiresEncoding)
+        {
+            Debug.Assert(idxOfFirstCharWhichRequiresEncoding >= 0);
+            Debug.Assert(idxOfFirstCharWhichRequiresEncoding < input.Length);
+
+            // The worst case encoding is 8 output chars per input char: [input] U+FFFF -> [output] "&#xFFFF;"
+            // We don't need to worry about astral code points since they consume *two* input chars to
+            // generate at most 10 output chars ("&#x10FFFF;"), which equates to 5 output per input.
+            int numCharsWhichMayRequireEncoding = input.Length - idxOfFirstCharWhichRequiresEncoding;
+            int sbCapacity = checked(idxOfFirstCharWhichRequiresEncoding + EncoderCommon.GetCapacityOfOutputStringBuilder(numCharsWhichMayRequireEncoding, worstCaseOutputCharsPerInputChar: 8));
+            Debug.Assert(sbCapacity >= input.Length);
+
+            // Allocate the StringBuilder with the first (known to not require encoding) part of the input string,
+            // then begin encoding from the last (potentially requiring encoding) part of the input string.
+            StringBuilder builder = new StringBuilder(input, 0, idxOfFirstCharWhichRequiresEncoding, sbCapacity);
+            fixed (char* pInput = input)
+            {
+                return HtmlEncodeImpl2(builder, &pInput[idxOfFirstCharWhichRequiresEncoding], (uint)numCharsWhichMayRequireEncoding);
+            }
+        }
+
+        private string HtmlEncodeImpl2(StringBuilder builder, char* input, uint charsRemaining)
+        {
+            while (charsRemaining != 0)
+            {
+                int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(input, endOfString: (charsRemaining == 1));
+                if (UnicodeHelpers.IsSupplementaryCodePoint(nextScalar))
+                {
+                    // Supplementary characters should always be encoded numerically.
+                    WriteScalarAsHtmlEncodedEntity(builder, (uint)nextScalar);
+
+                    // We consume two UTF-16 characters for a single supplementary character.
+                    input += 2;
+                    charsRemaining -= 2;
+                }
+                else
+                {
+                    // Otherwise, this was a BMP character.
+                    input++;
+                    charsRemaining--;
+                    char c = (char)nextScalar;
+                    if (IsCharacterAllowed(c))
+                    {
+                        builder.Append(c);
+                    }
+                    else
+                    {
+                        if (c == '<') { builder.Append("&lt;"); }
+                        else if (c == '>') { builder.Append("&gt;"); }
+                        else if (c == '&') { builder.Append("&amp;"); }
+                        else if (c == '\"') { builder.Append("&quot;"); }
```

Comment:
No - the 'apos' entity is undefined in HTML 4, so most encoders maintain compat by not emitting the entity.